hibernate-core/src/main/java/org/hibernate/loader/collection/DynamicBatchingCollectionInitializerBuilder.java

```diff
@@ -150,7 +155,17 @@ protected StringBuilder whereString(String alias, String[] columnNames, String s
 							return super.whereString( alias, columnNames, subselect, batchSize );
 						}
 
-						return StringHelper.buildBatchFetchRestrictionFragment( alias, columnNames, getFactory().getDialect() );
+						useArrayRestriction = BatchFetchStyle.ANY_ARRAY == factory.getSessionFactoryOptions()
+								.getBatchFetchStyle() &&
+								columnNames.length == 1;
```

Comment:
This will address the composite identifier use case, indeed, but we need to check the JDBC Driver to see if we could pass multiple arrays to match the available placeholders.

Reply:
Yes, frankly, I'm not sure it's doable but I think it's worth a shot to have a more generic approach.

Reply:
See https://www.postgresql.org/message-id/CAB%3DJe-HGL1LbhfH%3D2y67Dq_ysex4cj%3Djvwb7K34%2BqQM%2B6ntceQ%40mail.gmail.com

> 4) Note: array support is limited, so currently you might fail to pass
"array of composite type".